### ISSUE TITLE
Dynamic Links don't survive app install on iOS 11.3.1

### ISSUE THREAD

**Issue by richardliveraise** [firebase/firebase-ios-sdk#1244]
* Xcode version: 9.3
  * Firebase SDK version: 5.0.0
  * Firebase Component: Dynamic Links
  * Component version: 3.0.0

### Problem:
Dynamic links does not survive app install from app store on iOS 11.3.1
<bundle id>://google/link/?dismiss=1&is_weak_match=1
Note that this works fine on previous iOS versions (11.2 and below)

#### Steps to reproduce:
1. Click on Firebase Dynamic Links
2. Browser redirect to app store
3. Install from app store
4. Open the app
5. Dynamic link not passed

**Comment by richardliveraise** [comment]
@morganchen12 For sure.

```
- (BOOL)application:(UIApplication *)application
            openURL:(NSURL *)url
  sourceApplication:(NSString *)sourceApplication
         annotation:(id)annotation {

//the url here returns:
//google/link/?match_type=none&match_message=No pre-install link matched for this device.
or     
//google/link/?dismiss=1&is_weak_match=1

    FIRDynamicLink *dynamicLink = [[FIRDynamicLinks dynamicLinks] dynamicLinkFromCustomSchemeURL:url];

//dynamicLink.url is nil here.

}
```

Note that on iOS 11.2 it still works.

**Comment by richardliveraise** [comment]
@morganchen12 @dmandar Any update on this? Thanks!

**Comment by morganchen12** [comment]
No update yet, but we should be able to get to this this week.

**Comment by dmandar** [comment]
Can you confirm if your links are using the app preview page?

**Comment by richardliveraise** [comment]
@dmandar yes, we're skipping the app preview page. But this works prior to iOS 11.3.1

**Comment by morganchen12** [comment]
Mandar is currently investigating this, but hasn't come back with anything yet.

Internal bug filed at b/79657436.

**Comment by kbcodelogicx** [comment]
What about the issue ? If this issue persist in every latest OS version then what is the meaning of using Firebase Dynamic link ? And why they are not providing with any solutions to this when this issue is creating so much noise over the internet.

**Comment by dmandar** [comment]
@kbcodelogicx we are actively investigating this issue and will have an update very soon.

**Comment by richardliveraise** [comment]
@dmandar @morganchen12 this is fixed on iOS 11.4, but issue still persists on iOS 11.3

**Comment by morganchen12** [comment]
@richardliveraise is this iOS 11.3.1, or just iOS 11.3?

**Comment by richardliveraise** [comment]
@morganchen12 correction, issue is on 11.3.1, I don't have 11.3 device to confirm.

**Comment by dmandar** [comment]
@richardliveraise that's correct - the issue still persists only on that version - we have a bug fix in the works for it. Also, 11.4 or any other version (up to one digit after the decimal) should not exhibit the behavior.

**Comment by morganchen12** [comment]
It turns out some of the fixes are still in the process of being implemented. To clarify, this is a server issue, so you won't need an SDK update in order to receive the fix. I'll leave this issue open for the time being until all of the error cases are fixed.

**Comment by toco1001** [comment]
+1

**Comment by toco1001** [comment]
It's working on `iOS11.3`, but not `iOS11.3.1`

**Comment by fmkpro1984** [comment]
Hi @morganchen12 , I'm also having this issue and am troubled because iOS 11.3 has currently the majority of iOS userbase. No pressure but do we have any ETA on the server side fix? Thanks in advance...

**Comment by shali3** [comment]
Is it going to be a server side fix?

**Comment by morganchen12** [comment]
Yes.

**Comment by davidwernhart** [comment]
I am trying to implement Firebase Invites on iOS right now, and after an app install I am also always receiving myAppScheme://google/link/?dismiss=1&is_weak_match=1 instead of the deep link. Therefore FIRInvites cannot parse it and it does not work. Is this related to the issue with Dynamic Links? I have one iPhone 6 on iOS 11.4 and one iPhone 5 on 10.3.3 and it does not work on both of them.

Mind you that it does work on android (with the same firebase project) and on iOS when the app is already installed, so this problem is just occuring when the user is first installing the app after receiving an Invite.

**Comment by morganchen12** [comment]
@davidwernhart can you file a new issue and include the versions of the Firebase frameworks you're using?

**Comment by richardliveraise** [comment]
yes, this is fixed on iOS 11.3.1 . Thank you!!

**Comment by Younghun-Jeong** [comment]
i have 11.4.1 (iPhone7)

**Comment by MarsVard** [comment]
Is this issue still open? Our app launches deep links perfectly when it is already installed, but the deep links never persist after install.

When debugging we can see the deep link in the logs with a message saying "Deep link does not contain valid required params" but it does say "Link is uniquely matched for this device" at the same time...

Any suggestions?

**Comment by tamilvelraj** [comment]
Yes, it is not working in iOS 12.0 also.

**Comment by dmandar** [comment]
The backend issue has been fixed. Can you make sure that the target app has the same bundleID as that contained in your deep link (in case of a shortened link, you can go to the Firebase console to get the long version of the link). Also, do you have the (recommended) app preview page enabled?

**Comment by dmitry-solomadin** [comment]
Does not work on iOS 12.3
It does work when the preview page is enabled, though.

**Comment by Banck** [comment]
@dmandar Have the same issue, when preview page is disabled: No pre-install link matched for this device
With preview page enabled all is fine.
iOS version: 12.3

**Comment by averyvery** [comment]
I believe I'm having the same issue with custom domains + preview + deferred links on Firebase v5.20.2. I'm [building my links manually.](https://firebase.google.com/docs/dynamic-links/create-manually) When I do the following with the preview enabled...

- Uninstall the app
- Open a custom domain dynamic link with preview enabled, and tap "open"
- Install the app via Xcode

...openUrl in AppDelegate is correctly called with my link URL. However, adding `efr=1` to the link (skipping the preview page) and repeating the process, the openUrl is called with "[my bundle ID]://google/link/?dismiss=1&is_weak_match=1" instead. 

Reading the docs, I'm not sure if this is intentional or not — is the app preview _required_ to make deferred deep links work?

**Comment by morganchen12** [comment]
@averyvery it's not explicitly mentioned as required since DDL can work without it, but app preview does provide more accurate matching of the app installation to the device that tapped on the link. @dmandar can maybe elaborate more.

**Comment by Appkidd** [comment]
@dmandar When can we expect to see dynamic links working without the preview page and with a forced redirect? This disrupts our app's installation flow as it's one of the main ways users install our app.

**Comment by CyrilNb** [comment]
Same in our app. It's a shame that's not mentioned at all and we have to browse all internet to find this pot to finally have a clue about this bug. As many devs, I think the option to skip the app preview is a must-have, and the success of the dynamic link should not depend on it. Especially when you make this feature SKIPPABLE. 
I really encourage you to make this works without app preview cause it disrupts the installation flow  especially with a referral program.

**Comment by loobnin** [comment]
We also face same issues in our iOS apps. All works fine with *.page.link domains, but when we use a custom domain then Dynamic Links don't survive app install.

In that case it is always “appscheme://google/link/?request_ip_version=IP_V4&match_message=No pre-install link matched for this device.” URL that passed to the “application:openURL:sourceApplication:annotation:” method.

So we just can’t use our brand domains for Dynamic Links, and it harms effectiveness of our campaigns.

Is there any progress of fixing this issue? Or a workaround without enabling a Preview page?

**Comment by quetool** [comment]
@loobnin did you find any solution or made some progress on this issue?

**Comment by 0x7fs** [comment]
I hate to pile on with a +1 but we're still seeing this issue on iOS 13.x as well. Perhaps this issue should be reopened.

**Comment by morganchen12** [comment]
@0x7fs, can you open a new issue and try reproducing the issue with the [dynamic links quickstart](https://github.com/firebase/quickstart-ios/tree/master/dynamiclinks)?

**Comment by quetool** [comment]
@0x7fs could you check for `[UIApplicationDelegate application:openURL:options:]` function in your AppDelegate? In my case that was my issue, If you do have `openURL` delegate y your AppDelegate with nothing in it, somehow this is overriding the same method in `FirebaseDynamicLinksPlugin.m` so you need to delete that one in your `AppDelegate.m`, like I explain here https://github.com/FirebaseExtended/flutterfire/issues/100#issuecomment-541204778

So, this is `AppDelegate.m`
![Captura de Pantalla 2019-10-23 a la(s) 14 25 27](https://user-images.githubusercontent.com/14978705/67418423-202ef980-f5a1-11e9-9594-ea7fd952e9a4.png)

And this is `FirebaseDynamicLinksPlugin.m`
![Captura de Pantalla 2019-10-23 a la(s) 14 25 45](https://user-images.githubusercontent.com/14978705/67418486-39d04100-f5a1-11e9-8da5-7a74f149aa55.png)
